File: week_1/day-4-sliding-window/logest_substring_follow_up/longest_unique_substring_return_string.py
s = "pwwkew"  # INPUT

def longestSubstring(s):
    last_seen = {}
    left = 0
    max_len = 0
    start = 0

    for right in range(len(s)):
        if s[right] in last_seen and last_seen[s[right]] >= left:
            left = last_seen[s[right]] + 1

        last_seen[s[right]] = right

        if right - left + 1 > max_len:
            max_len = right - left + 1
            start = left

    return s[start:start + max_len]

print(longestSubstring(s))  # Output: "wke"

File: week_1/day-4-sliding-window/logest_substring_follow_up/longestUniqueSubstringReturnString.js
const s = "pwwkew"; // INPUT

function longestSubstring(s) {
    const map = new Map();
    let left = 0, maxLen = 0, start = 0;

    for (let right = 0; right < s.length; right++) {
        if (map.has(s[right]) && map.get(s[right]) >= left) {
            left = map.get(s[right]) + 1;
        }
        map.set(s[right], right);

        if (right - left + 1 > maxLen) {
            maxLen = right - left + 1;
            start = left;
        }
    }
    return s.substring(start, start + maxLen);
}

console.log(longestSubstring(s)); // "wke"
